### ISSUE TITLE
pgxpool: Callback for after pool initialization

### ISSUE THREAD

**Issue by jameshartig** [jackc/pgx#1660]
**Is your feature request related to a problem? Please describe.**
Currently there's no way to know when a pgxpool is done initializing and creating connections since this happens in a goroutine. The problem is that if you use the pool before it's done initializing then you'll end up with more connections than you intended.

```go
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

func main() {
	pool, err := pgxpool.New(context.Background(), "postgresql://postgres@127.0.0.1:5432?pool_min_conns=1&pool_max_conns=1")
	if err != nil {
		panic(err)
	}
	defer pool.Close()
	fmt.Println("before: ", pool.Stat().TotalConns())
	conn, err := pool.Acquire(context.Background()) // or any other method that uses the pool
	if err != nil {
		panic(err)
	}
	defer conn.Release()
	fmt.Println("after: ", pool.Stat().TotalConns())
	time.Sleep(time.Second)
	fmt.Println("after 2: ", pool.Stat().TotalConns())
}
```

Will print out:
```
before:  0
after:  2
after 2:  2
```

**Describe the solution you'd like**
Ideally the idle resource creation would see that there's already a constructing resource but that's very race-condition-prone. I went down this path for a while but there's not an easy way to 100% ensure that the pool doesn't get too many connections.

**Describe alternatives you've considered**
Alternatively, an `AfterInit()` function could be added to `pgxpool.Config` that would be called after the pool itself is done initializing the connections. This would allow anyone who cares about waiting for the pool to initialize to set that and wait.

Example:
```go
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

func main() {
	cfg, err := pgxpool.ParseConfig("postgresql://postgres@127.0.0.1:5432?pool_min_conns=1&pool_max_conns=1")
	if err != nil {
		panic(err)
	}
	doneCh := make(chan struct{})
	cfg.AfterInit = func() {
		close(doneCh)
	}
	pool, err := pgxpool.NewWithConfig(context.Background(), cfg)
	if err != nil {
		panic(err)
	}
	<-doneCh
	defer pool.Close()
	fmt.Println("before: ", pool.Stat().TotalConns())
	conn, err := pool.Acquire(context.Background())
	if err != nil {
		panic(err)
	}
	defer conn.Release()
	fmt.Println("after: ", pool.Stat().TotalConns())
	time.Sleep(time.Second)
	fmt.Println("after 2: ", pool.Stat().TotalConns())
}
```
Will print out:
```
before:  0
after:  1
after 2:  1
```

**Additional context**
Diff is:
```diff
diff --git a/pgxpool/pool.go b/pgxpool/pool.go
index f4c3a30c..64969cdf 100644
--- a/pgxpool/pool.go
+++ b/pgxpool/pool.go
@@ -123,6 +123,9 @@ type Config struct {
 	// BeforeClose is called right before a connection is closed and removed from the pool.
 	BeforeClose func(*pgx.Conn)
 
+	// AfterInit is called after the pool has created the initial connections.
+	AfterInit func()
+
 	// MaxConnLifetime is the duration since creation after which a connection will be automatically closed.
 	MaxConnLifetime time.Duration
 
@@ -260,6 +263,9 @@ func NewWithConfig(ctx context.Context, config *Config) (*Pool, error) {
 
 	go func() {
 		p.createIdleResources(ctx, int(p.minConns))
+		if config.AfterInit != nil {
+			config.AfterInit()
+		}
 		p.backgroundHealthCheck()
 	}()
 
```

**Comment by jackc** [comment]
> Currently there's no way to know when a pgxpool is done initializing and creating connections since this happens in a goroutine. The problem is that if you use the pool before it's done initializing then you'll end up with more connections than you intended.

If I understand correctly, the ultimate issue is that the max pool size can be briefly exceeded. I would consider this a bug. However, I would not consider it a bug if when MinConns=2 and MaxConns=10, and after a single Acquire the pool size is 3.

> Describe the solution you'd like
> Ideally the idle resource creation would see that there's already a constructing resource but that's very race-condition-prone. I went down this path for a while but there's not an easy way to 100% ensure that the pool doesn't get too many connections.

I would see this as the ideal solution as well. In particular, I do not think this issue is limited to new pools. The same thing could probably happen if a background health check happened at the same time as an Acquire.

I *think* there is a fairly simple change that would fix the issue. Though it does involve a change in documented behavior. The background creation uses https://pkg.go.dev/github.com/jackc/puddle/v2#Pool.CreateResource. It is documented as *not* checking the pool size. But I don't think there any reason for that (at least anymore). If that was changed it should prevent ever exceeding pool size.

> Describe alternatives you've considered
> Alternatively, an AfterInit() function could be added to pgxpool.Config that would be called after the pool itself is done initializing the connections. This would allow anyone who cares about waiting for the pool to initialize to set that and wait.

I'd prefer to have the underlying issue fixed if possible. Also, this could be mimicked fairly easily in the calling code by looping over a `Stat` call until min connections is reached.

**Comment by jameshartig** [comment]
@jackc I appreciate the quick reply and I apologize for the delay in my reply.

> If I understand correctly, the ultimate issue is that the max pool size can be briefly exceeded. I would consider this a bug. However, I would not consider it a bug if when MinConns=2 and MaxConns=10, and after a single Acquire the pool size is 3.
> The same thing could probably happen if a background health check happened at the same time as an Acquire.

Yes, I agree with both of those statements.

> I think there is a fairly simple change that would fix the issue. Though it does involve a change in documented behavior. The background creation uses https://pkg.go.dev/github.com/jackc/puddle/v2#Pool.CreateResource. It is documented as not checking the pool size. But I don't think there any reason for that (at least anymore). If that was changed it should prevent ever exceeding pool size.

To clarify: `CreateResource` would return `ErrNotAvailable` if the pool is at capacity and pgxpool would ignore that error when it creating idle resources.

**Comment by jackc** [comment]
@jameshartig See what you think of jackc/puddle@8101c0dff4beedfa171d281d65a7d5209b2952fc. I think that should prevent the maximum pool size being exceeded.

**Comment by jameshartig** [comment]
@jackc yep that looks good! Then we just need to handle that in pgxpool, right?

**Comment by jackc** [comment]
@jameshartig Yes, it does make sense to specifically catch and ignore that error for code clarity. However, background health check errors are never actually reported anywhere, so there won't be any user visible difference either way.

Should be fixed as of c513e2e435dca8acde76a4a0ed4856b9946b14e0.
